src/transform/qgstransform.cpp

```diff
@@ -0,0 +1,440 @@
+      return nullptr;
+    }
+  }
+
+  typedef QgsPythonUtils*( *inst )();
+  inst pythonlib_inst = reinterpret_cast< inst >( cast_to_fptr( pythonlib.resolve( "instance" ) ) );
+  if ( !pythonlib_inst )
+  {
+    //using stderr on purpose because we want end users to see this [TS]
+    std::cerr << "Couldn't resolve python support library's instance() symbol.";
+    return nullptr;
+  }
+
+  std::unique_ptr< QgsPythonUtils > pythonUtils( pythonlib_inst() );
+  if ( pythonUtils )
+  {
+    pythonUtils->initPython( nullptr, false );
+  }
+
+  return pythonUtils;
+}
+
+QgsProcessingExec::QgsProcessingExec()
+{
+
+}
+
+int QgsProcessingExec::run( const QStringList &args )
+{
+  if ( args.size() == 1 )
+  {
+    showUsage( args.at( 0 ) );
+    return 0;
+  }
+
+  // core providers
+  QgsApplication::processingRegistry()->addProvider( new QgsNativeAlgorithms( QgsApplication::processingRegistry() ) );
+#ifdef HAVE_3D
+  QgsApplication::processingRegistry()->addProvider( new Qgs3DAlgorithms( QgsApplication::processingRegistry() ) );
+#endif
+
+  // give Python plugins a chance to load providers
+  mPythonUtils = loadPythonSupport();
+  if ( !mPythonUtils )
+  {
+    QCoreApplication::exit( 1 );
+    return 1;
+  }
+  loadPlugins();
+
+  const QString command = args.at( 1 );
+  if ( command == QLatin1String( "plugins" ) )
+  {
+    listPlugins();
+    return 0;
+  }
+  else if ( command == QLatin1String( "list" ) )
+  {
+    listAlgorithms();
+    return 0;
+  }
+  else if ( command == QLatin1String( "help" ) )
+  {
+    if ( args.size() < 3 )
+    {
+      std::cerr << QStringLiteral( "Algorithm ID not specified\n" ).toLocal8Bit().constData();
+      return 1;
+    }
+
+    const QString algId = args.at( 2 );
+    return showAlgorithmHelp( algId );
+  }
+  else if ( command == QLatin1String( "run" ) )
+  {
+    if ( args.size() < 3 )
+    {
+      std::cerr << QStringLiteral( "Algorithm ID not specified\n" ).toLocal8Bit().constData();
+      return 1;
+    }
+
+    const QString algId = args.at( 2 );
+
+    // build parameter map
+    QVariantMap params;
+    for ( int i = 3; i < args.count(); i++ )
+    {
+      QString arg = args.at( i );
+      if ( arg.startsWith( QLatin1String( "--" ) ) )
+        arg = arg.mid( 2 );
+
+      const QStringList parts = arg.split( '=' );
+      if ( parts.count() >= 2 )
+      {
+        const QString name = parts.at( 0 );
+        const QString value = parts.mid( 1 ).join( '=' );
+        params.insert( name, value );
+      }
+      else
+      {
+        std::cerr << QStringLiteral( "Invalid parameter value %1. Parameter values must be of the form \"--NAME=VALUE\"\n" ).arg( arg ).toLocal8Bit().constData();
+        return 1;
+      }
+    }
+
+    return execute( algId, params );
+  }
+  else
+  {
+    std::cerr << QStringLiteral( "Command %1 not known!\n" ).arg( command ).toLocal8Bit().constData();
+  }
+  return 1;
+}
+
+void QgsProcessingExec::showUsage( const QString &appName )
```

Comment:
I would recommend to use QCommandLineParser as done here: https://github.com/qgis/QGIS/blob/master/src/server/qgis_mapserver.cpp#L137

Reply:
I just tried to do this, but it's not possible as QCommandLineParser requires knowledge up front of the "--" syntax options, whereas we need to adapt these based on the algorithm selected